### DISCUSSION TITLE
Read ballot formatting settings from an external file

### DISCUSSION THREAD

**Discussion by stratofax** [TrustTheVote-Project/BallotLab#101] [Ideas]
At some point, it would be useful to read the ballot formatting settings from an external file. The question is: what aspects of the ballot layout should be controllable via external settings? Some options include:

- Page size (letter, legal, A4, etc.).
- Font family, if we ever want to use something besides Helvetica.
- Font sizes -- might be useful to create a "large print" ballot for visually impared voters.
- Page layout, including margin and header size, column width.
- Toggling or modifying the voting instructions.

This is all just a start. What do you think?

**Comment by trustthevote** [comment]
And it is a good start! Later we will need to define a schema for it, and represent these and other things a JSON document in that schema. The path to this configuration file would be part of CLI, and anything currently in the CLI that is of this nature, should get moved to the config file. Expert mode use would be editing the config file with a JSON editor that enforces the schema. A later GUI would include a configurator UI to present the content of the config file, make changes with some syntax/semantic constraints, and write out a new config file. I really don't think that BallotLab itself would change in structure. I do imagine a Web UI that wraps BallotLab and calls it via CLI in response to GUI user input

**Reply by stratofax** [reply]
This all makes sense. The only caveat I'd add is this: currently, ballotmaker offers a "demo" option:

```
ballotmaker demo
```

This creates sample ballots using the default settings -- basically, in accordance with the EAC standards. One possible option would be for ballotmaker to write the default settings out to a JSON file, thus giving the experts something to start working with -- and a way to get back to a "know-good" configuration if they really mangle the settings file.

As discussed, the GUI could be a WebUI instead of a conventional Win/Mac/Linux app. As an example of a web GUI that can also run locally, I'm a big fan of the [fish shell](https://fishshell.com/), which uses a **local** web server to create a configuration GUI, which in turn writes out an updated config file. See [fish_config - start the web-based configuration interface — fish-shell 3.5.1 documentation](https://fishshell.com/docs/current/cmds/fish_config.html?highlight=web) for an excellent example of how a web UI can be used locally for CLI configuration.
